rpc: Add getindexinfo RPC

fjahr opened the pull request:
As I was playing with indices a I was missing an RPC that gives information about the active indices in the node. I think this can be helpful for many users, especially since there are some new index candidates coming up (#14053, #18000) that can give a quick overview without the user having to parse the logs.

Feature summary:
- Adds new RPC `getindexinfo` (placed in Util section)
- That RPC only lists the actively running indices
- For each index it gives the name, whether it is synced and up to which block height it is synced

davidinvest1 commented:
Please send me your phone 
    Il domenica 19 luglio 2020, 00:25:55 CEST, Fabian Jahr <notifications@github.com> ha scritto:  
 
 
As I was playing with indices a I was missing an RPC that gives information about the active indices in the node. I think this can be helpful for many users, especially since there are some new index candidates coming up (#14053, #18000) that can give a quick overview without the user having to parse the logs.

Feature summary:
   
   - Adds new RPC listindices (placed in Util section)
   - That RPC only lists the actively running indices
   - For each index it gives the name, whether it is synced and up to which block height it is synced

You can view, comment on, or merge this pull request online at:

  https://github.com/bitcoin/bitcoin/pull/19550

Commit Summary
   
   - rpc: Add listindices

File Changes
   
   -  M src/index/base.cpp (9) 
   -  M src/index/base.h (9) 
   -  M src/rpc/misc.cpp (46) 
   -  M test/functional/rpc_misc.py (10) 

Patch Links:
   
   - https://github.com/bitcoin/bitcoin/pull/19550.patch
   - https://github.com/bitcoin/bitcoin/pull/19550.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

promag reviewed (commented):
Concept ACK, why not? This can help the client determine whether the node is capable for it's needs.

Instead of returning an array, it could return an object `{ name: summary }`, or duplicate index names are allowed?

jonatack commented on src/rpc/misc.cpp:
```diff
@@ -4,6 +4,8 @@
 
 #include <httpserver.h>
+#include <index/txindex.h>
+#include <index/blockfilterindex.h>
```
nit: sort

jonatack commented on src/rpc/misc.cpp:
```diff
@@ -597,6 +599,49 @@ static UniValue echo(const JSONRPCRequest& request)
+                "\nReturns the available indices and their status.\n",
+                {},
+                RPCResult{
+                    RPCResult::Type::ARR, "", "",
```
Not sure it makes sense to do so here, but ISTM returning a JSON object is generally the preferred RPC result.

MarcoFalke replied:
The name should be unique, so it could be the key of the json object?

jonatack reviewed (commented):
Concept ACK, thanks for adding the test.

MarcoFalke commented on test/functional/rpc_misc.py:
```diff
@@ -11,6 +11,7 @@
     assert_equal,
     assert_greater_than,
     assert_greater_than_or_equal,
+    wait_until,
```
unrelated style nit: If you want you can cherry-pick the change to the test framework from https://github.com/bitcoin/bitcoin/pull/19134/files#diff-64721c5ee64d44f7114d6d0d2226db4d and then call self.wait_until below instead of the import.

MarcoFalke commented on src/rpc/misc.cpp:
```diff
@@ -611,6 +656,7 @@ static const CRPCCommand commands[] =
     { "util",               "getdescriptorinfo",      &getdescriptorinfo,      {"descriptor"} },
     { "util",               "verifymessage",          &verifymessage,          {"address","signature","message"} },
     { "util",               "signmessagewithprivkey", &signmessagewithprivkey, {"privkey","message"} },
+    { "util",               "listindices",            &listindices,            {} },
```
unrelated style-nit: Maybe use the "new" constructor that takes an RpcMethodFnType . See https://doxygen.bitcoincore.org/class_c_r_p_c_command.html#afa39f5bda9319f3b91c6b38bbc7c7434 and #19528

MarcoFalke commented on src/rpc/misc.cpp:
```diff
@@ -597,6 +599,49 @@ static UniValue echo(const JSONRPCRequest& request)
+    UniValue results(UniValue::VARR);
+
+    if (g_txindex) {
+        AddSummary(g_txindex->GetSummary(), results);
```
nit: Could avoid passing results around

```suggestion
        results.push_back(SummaryToJson(g_txindex->GetSummary()));
```

maflcko reviewed (approved):
review ACK, only style nits

theStack commented:
Concept ACK

fjahr commented:
Thanks for the reviews, I should have addressed all the comments:
- Now returning an object with the index names as keys (names are not enforced to be unique by the code at the moment I think but the names are used to distinguish the indices in the logs so they should be unique).
- Cherry-picked the commit that makes `wait_until` part of the TestFramework and used it
- Used new RPC constructor
- Not passing around results object anymore
- Fixed includes sorting

laanwj commented:
Tested ACK f301c86b77f027f350e646e19afe1d95a46bc24c
(tested with and without `-txindex` on regtest)

promag reviewed (commented):
Code review ACK f301c86b77f027f350e646e19afe1d95a46bc24c, needs release notes.

jonasschnelli commented:
I think this is a helpful addition for connecting applications.
Code Review utACK f301c86b77f027f350e646e19afe1d95a46bc24c.

jonatack commented:
This first commit is already merged in master.

Reviewing.

jonatack commented:
The first commit is already merged in master. The change was embedded into commit 12410b1feb80189061eb4a2b43421e53cbb758a8. Reviewing the main commit now.

fjahr commented:
Added const usage and research notes. Dropped first commit and rebased on master.

jonatack commented on src/rpc/misc.cpp:
```diff
@@ -597,6 +599,54 @@ static UniValue echo(const JSONRPCRequest& request)
+    return results;
+},
+    };
+}
```
Indentation, could clang-format `listindices()` if you retouch -- see details for an idea.

<details><summary><code>clang-format -i src/rpc/misc.cpp</code></summary><p>

```diff
 static RPCHelpMan listindices()
 {
-    return RPCHelpMan{"listindices",
-                "\nReturns the available indices and their status.\n",
-                {},
-                RPCResult{
-                    RPCResult::Type::OBJ, "", "", {
-                        {
-                            RPCResult::Type::OBJ, "name", "The name of the index",
-                            {
-                                {RPCResult::Type::BOOL, "synced", "Whether the index is synced or not"},
-                                {RPCResult::Type::NUM, "best_block_height", "The block height to which the index is synced"},
-                            }
-                        },
-                    },
-                },
-                RPCExamples{
-                    HelpExampleCli("listindices", "")
-                },
-                [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
-{
-    UniValue results(UniValue::VOBJ);
+    return RPCHelpMan{
+        "listindices",
+        "\nReturns the available indices and their status.\n",
+        {},
+        RPCResult{
+            RPCResult::Type::OBJ,
+            "",
+            "",
+            {
+                {RPCResult::Type::OBJ, "name", "The name of the index", {
.../...
+            },
+        },
+        RPCExamples{HelpExampleCli("listindices", "")},
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue {
+            UniValue results(UniValue::VOBJ);
 
-    if (g_txindex) {
-        results.pushKVs(SummaryToJSON(g_txindex->GetSummary()));
-    }
+            if (g_txindex) {
+                results.pushKVs(SummaryToJSON(g_txindex->GetSummary()));
+            }
 
-    ForEachBlockFilterIndex([&results](BlockFilterIndex& index) {
-        results.pushKVs(SummaryToJSON(index.GetSummary()));
-    });
+            ForEachBlockFilterIndex([&results](BlockFilterIndex& index) {
+                results.pushKVs(SummaryToJSON(index.GetSummary()));
+            });
 
-    return results;
-},
+            return results;
+        },
     };
 }
 ```
</p></details>

fjahr replied:
I have chosen this indentation to keep it consistent with https://github.com/bitcoin/bitcoin/pull/19528/files

jonatack commented on src/index/base.cpp:
```diff
@@ -319,3 +319,12 @@ void BaseIndex::Stop()
+
+const IndexSummary BaseIndex::GetSummary()
+{
+    IndexSummary summary;
```
Would `IndexSummary summary{};` provide further safety?

fjahr replied:
done

jonatack reviewed (commented):
ACK 3785cfd3c3 code review, debug build/tests, tested the RPC and its help.

fjahr commented on src/index/base.h:
```diff
@@ -13,6 +13,12 @@
+struct IndexSummary {
+    std::string name;
+    bool synced;
+    int best_block_height;
```
done

fjahr commented:
Also fixed initializations are proposed by @jonatack . I am leaving the indentation as is since it is consistent with #19528 and the suggested change from clang-format seem pretty different from our general formatting for RPCs so far.

fjahr commented on src/index/base.h:
```diff
@@ -106,6 +112,9 @@ class BaseIndex : public CValidationInterface
     void Stop();
+
+    /// Get a summary of the index and its state.
+    IndexSummary GetSummary();
```
done

jonatack commented:
Code review re-ACK 28b4cbfa5a5 per `git diff 3785cfd 28b4cbf`

Thanks, and sorry about that; I'm not sure the default initialization is useful in this case and in non-critical code, though I guess it can't hurt. When I saw [this catch today](https://github.com/bitcoin/bitcoin/pull/19534#discussion_r460407616) that I had missed in my own review, I resolved to reinforce my thinking about this.

fjahr commented on src/rpc/misc.cpp:
```diff
@@ -597,6 +599,54 @@ static UniValue echo(const JSONRPCRequest& request)
+        results.pushKVs(SummaryToJSON(g_txindex->GetSummary()));
+    }
+
+    ForEachBlockFilterIndex([&results](BlockFilterIndex& index) {
```
yep, done

fjahr commented on src/index/base.h:
```diff
@@ -106,6 +112,9 @@ class BaseIndex : public CValidationInterface
     void Stop();
+
+    /// Get a summary of the index and its state.
+    const IndexSummary GetSummary();
```
duh, done

fjahr commented:
Addressed review comments by @MarcoFalke

promag commented:
Code review ACK.

Feel free to squash commits.

I'd suggest to rename this RPC, but its probably too late. I was thinking `getindexstatus [index]`, with optional `index` parameter - if null/omitted returns all indices.

MarcoFalke commented:
ACK 8094e211d9af2bbe16d4de865aca5268bbc0f0f1

jonatack commented:
> I was thinking `getindexstatus [index]`, with optional `index` parameter - if null/omitted returns all indices.

Or `getindexinfo` for consistency and shorter to type.

jonatack commented:
> I was thinking `getindexstatus [index]`, with optional `index` parameter - if null/omitted returns all indices.

Or `getindexinfo` for consistency (e.g. `getpeerinfo`, `getnetworkinfo`) and shorter to type.

Per `bitcoin-cli help | grep list`, "list" RPCs deal with user-controlled data; "info" might make more sense here.

RPCs are hard to change once released, so it seems a good idea to get them right.

fjahr commented:
Thanks for the suggestions! Honestly, I was a bit on the fence on the naming of this RPC anyway and since there were two reviewers in favor of this renaming and none against, I have implemented it now as `getindexinfo`. It also takes an index name now as suggested. Sorry reviewers, most of the code has changed now.

Please note this trade-off in the code currently: Even when an index name (false or correct) is given it will still get the status for all/both indices internally. That could be easily fixed by matching the index name param against a fixed set of index names allowed. The upside of the current version is fewer LOC and that it has more flexibility in the sense that it will automatically work if a new (custom?) block filter index is introduced. I think this is ok since an internal optimization can still be done later and I am thinking of something like an `IndexMan` or so to encapsulate that logic for when we have more indices in the future. But of course, I can introduce some logic with hardcoded strings to handle this now if that is preferred.

I also went back on passing the result around, it seemed cleaner this way now. And squashed into one commit.

promag commented:
You may want to update PR title.

promag reviewed (commented):
Not sure if the format should be different if an index is specified. And if the specified index is unknown (either invalid or just because it's not available) then the resulting object would be `{}`. No strong opinion though.

jonatack commented:
> Not sure if the format should be different if an index is specified. And if the specified index is unknown (either invalid or just because it's not available) then the resulting object would be `{}`. No strong opinion though.

This was my first thought as well: if an index is specified, ISTM the output should give feedback on the index that it understood (by displaying it, the same as when no index is specified). I'll review properly tomorrow, so don't give this comment too much weight yet.

jonatack reviewed (commented):
ACK in general, with some suggestions.

Perhaps the param would best be named `index`, for what it is, rather than `name`. See the param names of the other MiscRPCCommands.

Perhaps it would be better to return the same output for one index, as for all indexes when only one index is active, thereby providing feedback that the RPC correctly understood the passed index.

Based on what most of the other PRs being merged are doing (see just-merged https://github.com/bitcoin/bitcoin/pull/19473/commits), I don't think you are required to squash this into one single commit as long as they are hygienic. The more frequent pattern for this change would seem to be one commit for the change, one for the tests, and one for the release note.

fjahr committed:
rpc: Add getindexinfo RPC

fjahr committed:
test: Add tests for getindexinfo RPC

fjahr committed:
doc: Add release notes for getindexinfo RPC

fjahr commented:
Pushed the changes as requested. I understand now the name parameter was more thought of like a filter so I documented it as such. I also renamed it to `index_name` which is consistent with `wallet_name` which seemed the most comparable.

fjahr commented on doc/release-notes.md:
```diff
@@ -115,6 +115,10 @@ Changes to Wallet or GUI related RPCs can be found in the GUI or Wallet section
 New RPCs
 --------
 
+- The `getindexinfo` RPC by default lists the actively running indices of the
```
done

jonatack commented on src/rpc/misc.cpp:
```diff
@@ -597,6 +599,62 @@ static UniValue echo(const JSONRPCRequest& request)
+    return RPCHelpMan{"getindexinfo",
+                "\nReturns the status of one or all available indices currently running in the node.\n",
+                {
+                    {"index_name", RPCArg::Type::STR, /* default */ "", "Filter results for an index with a specific name."},
```
Would (something like) this be better?
```suggestion
                    {"index_name", RPCArg::Type::STR, RPCArg::Optional::OMITTED_NAMED_ARG, "Filter results for an index with a specific name."},
```

It would change the help from
```
Arguments:
1. index_name    (string, optional, default=) Filter results for an index with a specific name.
```
to
```
Arguments:
1. index_name    (string, optional) Filter results for an index with a specific name.
```

jonatack replied:
(I should have given feedback about the "default=" earlier, but somehow kept forgetting to add it to the review)

jonatack commented on src/rpc/misc.cpp:
```diff
@@ -597,6 +599,62 @@ static UniValue echo(const JSONRPCRequest& request)
+    UniValue ret_summary(UniValue::VOBJ);
+    if (index_name != "" && index_name != summary.name) {
+        return ret_summary;
+    }
```
nit: not sure if `empty()` is more idiomatic
```diff
-    if (index_name != "" && index_name != summary.name) {
-        return ret_summary;
-    }
+    if (!index_name.empty() && index_name != summary.name) return ret_summary;
```

jonatack reviewed (commented):
ACK dc888cdb07 with a question for the help

fjahr commented:
Addressed review comments by @jonatack

jonatack reviewed (approved):
Code review re-ACK 47a5372 per `git diff dc888cd 47a5372` :racehorse:

jonatack reviewed (commented):
Code review re-ACK 124e1ee per `git range-diff a57af89 47a5372 124e1ee` no change since my last re-ACK, rebase only

laanwj commented:
Re-ACK 124e1ee1343f8bfb3748393ced9debdbdee60d3b
